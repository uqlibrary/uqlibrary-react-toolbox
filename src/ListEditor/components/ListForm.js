import React, {Component} from 'react';
import PropTypes from 'prop-types';
import TextField from 'material-ui/TextField';
import RaisedButton from 'material-ui/RaisedButton';

export default class ListForm extends Component {

    static propTypes = {
        onAdd: PropTypes.func.isRequired,
        isValid: PropTypes.func,
        locale: PropTypes.object,
        disabled: PropTypes.bool
    };

    static defaultProps = {
        isValid: () => { return ''; },
        locale: {
            inputFieldLabel: 'Item name',
            inputFieldHint: 'Please type the item name',
            addButtonLabel: 'Add'
        }
    }

    constructor(props) {
        super(props);
        this.state = {
            itemName: ''
        };
    }

    addItem = (event) => {
        // add item if user hits 'enter' key on input field
        if(this.props.disabled
            || this.props.isValid(this.state.itemName) !== ''
            || (event && event.key && (event.key !== 'Enter' || this.state.itemName.length === 0))) {
            return;
        }

        // pass on the selected item
        this.props.onAdd(this.state.itemName);

        // reset internal state
        this.setState({
            itemName: ''
        });

        // move focus to name as published text field after item was added
        if (this.refs.itemName) this.refs.itemName.focus();
    }

    onNameChanged = (event, newValue) => {
        this.setState({
            itemName: newValue
        });
    }

    render() {
        return (
            <div className="columns">
                <div className="column">
                    <TextField
                        fullWidth
                        ref="itemName"
                        floatingLabelText={this.props.locale.inputFieldLabel}
                        hintText={this.props.locale.inputFieldHint}
                        value={this.state.itemName}
                        onChange={this.onNameChanged}
                        onKeyPress={this.addItem}
                        errorText={this.props.isValid(this.state.itemName)}
                        disabled={this.props.disabled} />
                </div>
                <div className="column is-narrow">
                    <RaisedButton
                        className="is-mui-spacing-button"
                        fullWidth
                        primary
                        label={this.props.locale.addButtonLabel}
                        disabled={this.props.disabled || this.props.isValid(this.state.itemName) !== '' || this.state.itemName.trim().length === 0}
                        onClick={this.addItem} />
                </div>
            </div>
        );
    }
}
