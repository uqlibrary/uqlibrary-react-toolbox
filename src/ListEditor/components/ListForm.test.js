jest.dontMock('./ListForm');

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import ListForm from './ListForm';
import injectTapEventPlugin from 'react-tap-event-plugin';

beforeAll(() => {
    injectTapEventPlugin();
});

function setup({onAdd, isValid, disabled}){

    const props = {
        onAdd: onAdd || jest.fn(), // : PropTypes.func.isRequired,
        isValid: isValid || jest.fn(() => ('')), // PropTypes.func,
        disabled // : PropTypes.bool
        //locale, // : PropTypes.object,
    };
    return shallow(<ListForm {...props} />);
}

describe('ListForm tests ', () => {
    it('rendering active form', () => {
        const wrapper = setup({ });
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('rendering disabled form', () => {
        const wrapper = setup({ disabled: true });
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('adding item method is called', () => {
        const testMethod = jest.fn();
        const wrapper = setup({ onAdd: testMethod });
        wrapper.setState({itemName: 'one'});
        wrapper.instance().addItem({});
        expect(testMethod).toBeCalled;
    });

    it('adding item method is not called on disabled form', () => {
        const testMethod = jest.fn();
        const wrapper = setup({ disabled: true, onAdd: testMethod });
        wrapper.setState({itemName: 'one'});
        wrapper.instance().addItem({});
        expect(testMethod).not.toBeCalled;
    });

    it('setting state', () => {
        const wrapper = setup({ });
        expect(wrapper.state().itemName).toBeFalsy();
        wrapper.instance().onNameChanged({}, 'one');
        expect(wrapper.state().itemName).toEqual('one');
    });
});
