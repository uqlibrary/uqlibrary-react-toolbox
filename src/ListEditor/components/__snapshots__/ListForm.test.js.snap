// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListForm tests  rendering active form 1`] = `
<div
  className="columns"
>
  <div
    className="column"
  >
    <TextField
      disabled={false}
      errorText=""
      floatingLabelFixed={false}
      floatingLabelText="Item name"
      fullWidth={true}
      hintText="Please type the item name"
      multiLine={false}
      onChange={[Function]}
      onKeyPress={[Function]}
      rows={1}
      type="text"
      underlineShow={true}
      value=""
    />
  </div>
  <div
    className="column is-narrow"
  >
    <RaisedButton
      className="is-mui-spacing-button"
      disabled={true}
      fullWidth={true}
      label="Add"
      labelPosition="after"
      onClick={[Function]}
      primary={true}
      secondary={false}
    />
  </div>
</div>
`;

exports[`ListForm tests  rendering disabled form 1`] = `
<div
  className="columns"
>
  <div
    className="column"
  >
    <TextField
      disabled={true}
      errorText=""
      floatingLabelFixed={false}
      floatingLabelText="Item name"
      fullWidth={true}
      hintText="Please type the item name"
      multiLine={false}
      onChange={[Function]}
      onKeyPress={[Function]}
      rows={1}
      type="text"
      underlineShow={true}
      value=""
    />
  </div>
  <div
    className="column is-narrow"
  >
    <RaisedButton
      className="is-mui-spacing-button"
      disabled={true}
      fullWidth={true}
      label="Add"
      labelPosition="after"
      onClick={[Function]}
      primary={true}
      secondary={false}
    />
  </div>
</div>
`;
